core/src/main/scala/org/http4s/client/jdkhttpclient/JdkHttpClient.scala

```diff
@@ -58,38 +60,166 @@ object JdkHttpClient {
+    // the body of the fromPublisher method in fs2.
+    //
+    // https://github.com/typelevel/fs2/blob/v2.5.0/reactive-streams/src/main/scala/fs2/interop/reactivestreams/package.scala#L55
+    //
+    // In the release section of bracket on the response, we also attempt to
+    // acquire the single permit from the Semaphore. If we are successfully
+    // able to do so, that means that Publisher was never subscribed to,
+    // either due to an error or more likely because the calling code didn't
+    // care about the body of the request. In this case we subscribe to the
+    // body and then immediately cancel the subscription, freeing the
+    // resources. If the Semaphore has already been drained, we do nothing.
+    //
+    // There are a couple items worth giving special attention to here.
+    //
+    // * It is important that we attach the finalizer which can run
+    // AlwaysCancelingSubscriber as soon as we run the effect to trigger the
+    // response. We _could_ attach it later, e.g. when we attach the body
+    // interruption signal, but this happens after a number of other side
+    // effects are run. If anything triggers abnormal termination _after_ we
+    // have the HttpResponse, but _before_ we've attached this finalizer, then
+    // we will have a resource leak.
+    //
+    // * Interrupting the response body stream alone will not trigger
+    // subscription and cancellation. Subscription only happens after someone
+    // attempts to pull from the Stream, which doesn't happen in cases where
+    // the body is discarded. Interrupting the fs2.Stream with a second
+    // Resource is still required to cleanup the fs2.Stream scopes, whether or
+    // not the Publisher was ever subscribed to.
     def convertResponse(
-        res: HttpResponse[Flow.Publisher[util.List[ByteBuffer]]]
+        responseF: F[HttpResponse[Flow.Publisher[util.List[ByteBuffer]]]]
     ): Resource[F, Response[F]] =
-      Resource(
-        (F.fromEither(Status.fromInt(res.statusCode)), SignallingRef[F, Boolean](false)).mapN {
-          case (status, signal) =>
-            Response(
-              status = status,
-              headers = Headers(res.headers.map.asScala.flatMap { case (k, vs) =>
-                vs.asScala.map(Header(k, _))
-              }.toList),
-              httpVersion = res.version match {
-                case HttpClient.Version.HTTP_1_1 => HttpVersion.`HTTP/1.1`
-                case HttpClient.Version.HTTP_2 => HttpVersion.`HTTP/2.0`
-              },
-              body = FlowAdapters
-                .toPublisher(res.body)
-                .toStream[F]
-                .interruptWhen(signal)
-                .flatMap(bs => Stream.fromIterator(bs.iterator.asScala.map(Chunk.byteBuffer)))
-                .flatMap(Stream.chunk)
-            ) -> signal.set(true)
+      Resource
+        .make(
+          (Semaphore[F](1L), responseF).tupled
```

Comment:
While using a semaphore with 1 permit works perfectly fine, I think it is somewhat nicer to use a `TryableDeferred` here, as this emphasizes that we only have one state transition (from "incomplete" to "complete".

Feel free to use/adapt [this diff](https://gist.github.com/amesgen/4cd0110da0271acdef96685edb1dc97d).